src/runtime/hexagon_remote/halide_remote.cpp

```diff
@@ -93,15 +93,32 @@ int halide_hexagon_remote_load_library(const char *soname, int sonameLen,
 }
 
 volatile int power_ref_count = 0;
+static volatile void *power_context = NULL;
+
+static void *get_HAP_power_context() {
+    if (power_context == NULL) {
+        power_context = (void *)malloc(1);
```

Comment:
Why not just pass an address of a global value to the function instead of a dynamic allocation?

Does it even need to be a valid allocation? (Currently uninitialized memory is being passed.)

Reply:
The power team says that the pointer returned from malloc will be unique over all processes.  The value does not need to be initialized.  The context ID is the address returned from malloc.

Reply:
So the goal here is just to get a unique ID? Can you comment something to that affect?

I also think there's probably a better way to get this. Is there a function to get a process ID on Hexagon, like getpid()?

It would be nice to avoid the need to free this thing.

Reply:
Also, I find it rather suspicious that malloc is guaranteed to be unique over all processes. That suggests processes are sharing an address space, which I think (I hope...) is not the case?

Reply:
Yes, we were surprised as well by that claim....but haven't tried to prove/disprove it ourselves.  Could easily be that the malloc produces a "unique-enough" context ID over all processes that power/voting works better...as opposed to all processes just using a NULL context ID.  I am not familiar with what goes on in the power/voting with the context ID....just that the newer devices refuse to dispatch is the context ID is NULL...and this is the recommended solution from the power team.

Reply:
...and the free isn't explicitly called at the moment....it is just happening on process exit for now.  For a long running process with multiple dispatches, I really didn't see a need to free the power context after a dispatch and get a new one on the next....esp. since this approach get/free isn't sending any information to the power controls.

Reply:
We could leave the power context change out of this PR if you like....and revisit it separately.  It came up at the same time as getting the DMA-BUF changes working on Android-S/12...and since it was also in the Halide runtime, I figured it would be good to get both of these changes in at once so they were there from day 1 on official release of Android-S/12.   The non-NULL power context is also only needed for devices newer than SM8350 (avoids breaking any currently deployed apps)...another reason we could leave it out for you.

Reply:
If free_HAP_power_context is dead code, then get rid of it?

In general I think this whole thing is pretty sketchy:
- Why does HAP_power_set need a system wide unique identifier? It should track the process votes come from itself if it cares about that.
- Why is the argument a pointer if it isn't going to be dereferenced?
- Getting a unique ID with malloc is both unlikely to actually work and also kind of sloppy (never freed, calling malloc just to get an identifier is weird).

IMHO the best thing here is to get a proper process unique ID somehow, and cast it to a pointer.

Reply:
OK, I just went through the latest internal docs from the power team and found:
- it only has to be unique within the process
- they are still recommending the malloc(1)/free for creating/destroying a context
- the context is just an identifier....even though it is a void*, it is never stored into....they even say you could even just use "1" (non-NULL), but that would not be unique within the process

I have a feeling that they used (void *) because they are using malloc to create process-wide unique IDs.

The reason for this change to no longer allow NULL power context is:
"If there are multiple clients running in the same process(e.g., camera and ML clients) and both of them using NULL context for voting purpose. Camera might vote using NULL and ML might remove vote using same NULL context. fastrpc wouldn't know to remove (or) retain vote as both clinets are using same NULL context."

Reply:
In that case, I suggest just passing the address of a global variable in our runtime. That should give equivalent unique-ness to what you have here (a global variable assigned by malloc).

Reply:
There is a new answer in the FAQ, however, that wasn't there before that makes me realize that we do need to still find a point at which we can say we are done with a context (specifically see the mention of HAP_power_destroy_client):

```
Is there any limit on number of contexts (or) clients contexts that can be created?

Yes there is limit on number of client context that can be created. This limit is coming from MMPM
to create voting client context. There is a limit of 192 power clients at a time by MMPM on the DSP.
Old client contexts need to be destroyed(HAP_power_destroy_client) to create new client contexts.
```

The HAP_power_destroy_client function doesn't exist in SDK 3.3.3 incs/HAP_power.h....was added in SDK 3.5.1 and is actually supposed to be called through a wrapper HAP_power_destroy which checks (weak) for HAP_power_destroy_client exists before calling it.

Reply:
Not sure if halide_hexagon_remote_power_hvx_off() is the right place (since we would probably need to revote when we turn hvx back on).   I'm thinking we need to register a function via atexit() to call HAP_power_destroy() (if the power context is non-NULL) as the process exits.  What do you think?

Reply:
Do we really need to call HAP_power_destroy_client at process exit? It's not going to get cleaned up anyways? What happens if the app crashes?

Reply:
Good question.   It is possible that the process exiting is sufficient for fastrpc to know we are done.  We may not even need to call HAP_power_destroy_client if fastrpc will itself releases the associated clients/contexts.  I'll ask the power team.

Reply:
Answer:
"It is ideal to always call HAP_power_destroy In your process tear down process.
Once the process exits(normal or exception) fastrpc teardown will take of releasing all client contexts. It shouldn’t leak once the DSP User PD exits completely."

Reply:
Pushed up a new change without the malloc/free and not calling HAP_power_destroy on the context ID.

Reply:
...and if we aren't ever going to free/change the power_context ID....I just pushed up a simplifying change that just returns the global power_context, doesn't store it.  This should be fine, but I'm rerunning new tests on this locally (on all old & new devices/OSes).  Will report back later.

Reply:
Testing looks good on this latest change.  Ran fine both on older and newer devices/OSes...including the newest device that requires a non-NULL context.

Reply:
Would you like me to add some of the above explanation as comments here?